File: src/engine/masker.py
"""
Masker: replaces detected spans with typed positional placeholders.

Placeholder format (D3): [ENTITY_TYPE_N]
  - Typed counters beat UUIDs: better tokenization, model can reason about relationships.
  - Counters are per-type within a request.
  - Consistent within a request: same original value → same placeholder.

Returns:
  - masked_text: the text with all spans replaced.
  - placeholder_map: {placeholder: original_value} — given to Vault.
"""

from __future__ import annotations

from collections import defaultdict

from .entities import DetectedSpan, EntityType


class Masker:
    def mask(
        self,
        text: str,
        spans: list[DetectedSpan],
    ) -> tuple[str, dict[str, str]]:
        """
        Replace spans with placeholders.  Consistent: same value → same placeholder.

        Returns:
            (masked_text, {placeholder: original_value})
        """
        if not spans:
            return text, {}

        counters: dict[EntityType, int] = defaultdict(int)
        value_to_placeholder: dict[str, str] = {}
        placeholder_to_value: dict[str, str] = {}

        # Sort by position descending so we can replace right-to-left without
        # invalidating earlier offsets.
        sorted_spans = sorted(spans, key=lambda s: s.start, reverse=True)

        # Pre-assign placeholders (forward pass for consistent numbering).
        forward_spans = sorted(spans, key=lambda s: s.start)
        for span in forward_spans:
            if span.original_value not in value_to_placeholder:
                counters[span.entity_type] += 1
                placeholder = f"[{span.entity_type.value}_{counters[span.entity_type]}]"
                value_to_placeholder[span.original_value] = placeholder
                placeholder_to_value[placeholder] = span.original_value

        # Apply replacements right-to-left.
        chars = list(text)
        for span in sorted_spans:
            placeholder = value_to_placeholder[span.original_value]
            chars[span.start : span.end] = list(placeholder)

        return "".join(chars), placeholder_to_value

File: crates/engine-core/src/masker.rs
/*!
Masker: replaces detected spans with typed positional placeholders (D3).

Format: `[ENTITY_TYPE_N]`  — consistent within a request (same original value
→ same placeholder), typed counters per entity type.

Operates right-to-left to preserve byte offsets of earlier spans.
*/

use std::collections::HashMap;

use crate::entities::{DetectedSpan, EntityType};

pub struct Masker;

impl Masker {
    pub fn mask(&self, text: &str, spans: &[DetectedSpan]) -> (String, HashMap<String, String>) {
        if spans.is_empty() {
            return (text.to_string(), HashMap::new());
        }

        // Forward pass: assign placeholders in first-seen order.
        let mut counters: HashMap<EntityType, usize> = HashMap::new();
        let mut value_to_placeholder: HashMap<String, String> = HashMap::new();
        let mut placeholder_to_value: HashMap<String, String> = HashMap::new();

        let mut sorted_fwd: Vec<&DetectedSpan> = spans.iter().collect();
        sorted_fwd.sort_by_key(|s| s.start);

        for span in &sorted_fwd {
            if !value_to_placeholder.contains_key(&span.original_value) {
                let counter = counters.entry(span.entity_type).or_insert(0);
                *counter += 1;
                let placeholder = format!("[{}_{}]", span.entity_type.as_str(), counter);
                value_to_placeholder.insert(span.original_value.clone(), placeholder.clone());
                placeholder_to_value.insert(placeholder, span.original_value.clone());
            }
        }

        // Backward pass: replace spans right-to-left.
        let mut sorted_bwd = sorted_fwd;
        sorted_bwd.sort_by_key(|s| std::cmp::Reverse(s.start));

        let mut chars: Vec<u8> = text.as_bytes().to_vec();
        for span in sorted_bwd {
            let placeholder = value_to_placeholder[&span.original_value].as_bytes();
            chars.splice(span.start..span.end, placeholder.iter().copied());
        }

        let masked = String::from_utf8(chars).expect("masked text is valid UTF-8");
        (masked, placeholder_to_value)
    }
}
